var/spack/repos/spack_repo/builtin/packages/llvm/package.py

```diff
@@ -805,26 +805,48 @@ def determine_version(cls, exe):
     def determine_variants(cls, exes, version_str):
         # Do not need to reuse more general logic from CompilerPackage
         # because LLVM has kindly named compilers
-        variants, compilers = {"+clang"}, {}
-        lld_found, lldb_found = False, False
-        for exe in sorted(exes, key=len):
-            name = os.path.basename(exe)
-            if "clang++" in name:
-                compilers.setdefault("cxx", exe)
-            elif "clang" in name:
-                compilers.setdefault("c", exe)
-            elif "flang" in name and "fortran" not in compilers:
-                variants.add("+flang")
-                compilers.setdefault("fortran", exe)
-            elif "ld.lld" in name:
-                lld_found = True
-            elif "lldb" in name:
-                lldb_found = True
-
-        variants.add("+lld" if lld_found else "~lld")
-        variants.add("+lldb" if lldb_found else "~lldb")
-
-        return "".join(sorted(variants)), {"compilers": compilers}
+
+        # Map between exectuable name, variant name, and compiler language.
+        # The ordering of this list is the ordering of the returned variants.
+        exe_variant_lang = [
+            ("clang++", "clang", "cxx"),
+            ("clang", "clang", "c"),
+            ("flang", "flang", "fortran"),
+            # TODO: gold?
+            ("ld.lld", "lld", None),
+            ("lldb", "lldb", None),
+        ]
+
+        variants = set()
+        compilers = {}
+
+        # Prefer shorter pathnames by sorting with len and using setdefault
+        for exe_path in sorted(exes, key=len):
+            name = os.path.basename(exe_path)
+            for exe, var, lang in exe_variant_lang:
+                # XXX: should this be `name.startswith(exe)`?
+                if exe in name:
```

Comment:
This is true to the previous implementation, but I'm not sure if we should process a hypothetical `/usr/bin/prolificlanguage` as `clang`....

Reply:
I think startswith makes sense. Suffixes exist for sure, I haven't really seen any prefix other than `amdclang`, but that's considered a different package in Spack.

Reply:
Oh, amdclang++ is probably the reason, since this code is shared.

Reply:
Aha, that makes sense.